### PR TITLE
Tests for provider deployment

### PR DESCRIPTION
**Pull request by bindrad**
Signed-off-by: bindrad <dbindra@redhat.com>

### CODE REVIEW

File: controllers/managedocs_controller_test.go [hunk 1]
```diff
@@ -640,7 +651,7 @@ var _ = Describe("ManagedOCS controller", func() {
 			})
 			When("there is a valid size in the add-on parameter secret", func() {
```
Comment (nb-ohad) [inline controllers/managedocs_controller_test.go:652]: Why does this case have 3 different its that validate the same "should create reconciled resources" with a different skip?

To me is seems like what we want is a single "it" with an internal switch statement for the difference details

File: controllers/managedocs_controller_test.go [hunk 2]
```diff
@@ -1139,8 +1170,8 @@ var _ = Describe("ManagedOCS controller", func() {
 			})
 			When("a Grafana datasources secret exists in the openshift-monitoring namespace", func() {
 				It("should create k8sMetricsServiceMonitorAuthSecret in primary namespace", func() {
-					grafanaSecret := grafanaDatasourceSecretTemplate.DeepCopy()
-					Expect(k8sClient.Create(ctx, grafanaSecret)).Should(Succeed())
+					secret := grafanaDatasourceSecretTemplate.DeepCopy()
+					Expect(k8sClient.Get(ctx, utils.GetResourceKey(secret), secret)).Should(Succeed())
```
Comment (nb-ohad) [inline controllers/managedocs_controller_test.go:1174]: if you assume the secret is already created you should not test for it, 
Expect is used to validate test results, not assumptions

File: controllers/suite_test.go [hunk 3]
```diff
@@ -177,6 +177,22 @@ var _ = BeforeSuite(func() {
...
+		}
+		grafanaDatasources.Datasources[0].BasicAuthPassword = "password"
+		grafanaDatasources.Datasources[0].BasicAuthUser = "internal"
+		grafanaDatasourceSecret := &corev1.Secret{}
+		grafanaDatasourceSecret.Name = testGrafanaFederateSecretName
+		grafanaDatasourceSecret.Namespace = testOpenshiftMonitoringNamespace
+		grafanaDatasourceSecret.Data = map[string][]byte{
+			"prometheus.yaml": utils.ToJsonOrDie(grafanaDatasources),
+		}
+		Expect(k8sClient.Create(ctx, grafanaDatasourceSecret)).ShouldNot(HaveOccurred())
```
Comment (nb-ohad) [inline controllers/suite_test.go:194]: Why was this moved here?

File: controllers/managedocs_controller_test.go [hunk 4]
```diff
@@ -676,6 +687,26 @@ var _ = Describe("ManagedOCS controller", func() {
...
+					}
+					// Create a valid add-on parameters secret
+					secret := addonParamsSecretTemplate.DeepCopy()
+					secret.Data["size"] = []byte("1")
+					secret.Data["enable-mcg"] = []byte("false")
+					secret.Data["onboarding-validation-key"] = []byte("test-validation-key")
+					Expect(k8sClient.Create(ctx, secret)).Should(Succeed())
+
+					By("Creating a storagecluster resource")
+					utils.WaitForResource(k8sClient, ctx, scTemplate.DeepCopy(), timeout, interval)
```
Comment (nb-ohad) [inline controllers/managedocs_controller_test.go:705]: We have different deployment types now so we need to validate the content as well. 
I think a deep equal on a desired outcome would be good

File: controllers/managedocs_controller_test.go [hunk 5]
```diff
@@ -536,6 +531,22 @@ var _ = Describe("ManagedOCS controller", func() {
 					Expect(k8sClient.Delete(ctx, secret)).Should(Succeed())
 				})
 			})
+			When("there is no onboarding-validation-key field in the add-on parameters secret", func() {
```
Comment (nb-ohad) [inline controllers/managedocs_controller_test.go:534]: Where is the opposite test that verifies successful creation with correct content?

**Review by nb-ohad — CHANGES_REQUESTED** [review]

File: controllers/suite_test.go (hunk 3 above)
Comment (bindrad) [reply controllers/suite_test.go:194]: This resource should by default exist in the monitoring namespace and it shouldn't be created while the test is running. So added it to BeforeSuite which ensures that the resource exists in the monitoring namespace.

Comment (nb-ohad) [reply controllers/suite_test.go:194]: How did it work before?

Comment (bindrad) [reply controllers/suite_test.go:194]: Before we were creating the secret in monitoring namespace in managedocs_controller_test.go file just before we check for k8sMetricsServiceMonitorAuthSecret
https://github.com/red-hat-storage/ocs-osd-deployer/blob/main/controllers/managedocs_controller_test.go#L1142-L1143

File: controllers/managedocs_controller_test.go [hunk 6]
```diff
@@ -1594,6 +1610,15 @@ var _ = Describe("ManagedOCS controller", func() {
...
 			})
+			When("there is onboarding-validation-key field in the add-on parameters secret", func() {
+				It("should create onboarding-ticket-key secret", func() {
+					if testReconciler.DeploymentType != providerDeploymentType {
+						Skip(fmt.Sprintf("Skipping the test as it is not required by %v", testReconciler.DeploymentType))
+					}
+					secret := onboardingValidationKeySecretTemplate.DeepCopy()
+					Expect(k8sClient.Get(ctx, utils.GetResourceKey(secret), secret)).Should(Succeed())
+				})
+			})
```
Comment (nb-ohad) [inline controllers/managedocs_controller_test.go:1621]: This test does not make sense? 
The secret is created whenever we deploy with the correct params, as such this needs to be checked together with the other resources that are created in that test.

Here you have no set up, so it cannot be a `when`

File: controllers/suite_test.go (hunk 3 above)
Comment (nb-ohad) [reply controllers/suite_test.go:194]: The old way makes more sense as it is the trigger for the test. 
Right now you do not actually have a test as you do not have a trigger for it. 

Let's revert.

**Review by nb-ohad — CHANGES_REQUESTED** [review]

File: controllers/managedocs_controller_test.go [hunk 7]
```diff
@@ -640,13 +646,18 @@ var _ = Describe("ManagedOCS controller", func() {
 			})
 			When("there is a valid size in the add-on parameter secret", func() {
 				It("should create reconciled resources", func() {
-					if testReconciler.DeploymentType == consumerDeploymentType {
-						Skip(fmt.Sprintf("Skipping the test as it is not required by %v", testReconciler.DeploymentType))
-					}
+
```
Comment (leelavg) [inline controllers/managedocs_controller_test.go:649]: - shouldn't this test be skipped for converged?

Comment (bindrad) [reply controllers/managedocs_controller_test.go:649]: No, this test should be for all the deployment types. The resources that gets created might be different depending on deployment types. And the storageCluster spec is different depending on deployment type.

**Review by leelavg — APPROVED** [review]
/lgtm

File: controllers/managedocs_controller_test.go [hunk 8]
```diff
@@ -640,44 +646,69 @@ var _ = Describe("ManagedOCS controller", func() {
...
 					By("Creating a prometheus resource")
 					utils.WaitForResource(k8sClient, ctx, promTemplate.DeepCopy(), timeout, interval)
 
 					By("Creating an alertmanager resource")
 					utils.WaitForResource(k8sClient, ctx, amTemplate.DeepCopy(), timeout, interval)
+
+					if testReconciler.DeploymentType == providerDeploymentType {
+						By("Creating a onboarding validation key secret resource")
+						utils.WaitForResource(k8sClient, ctx, onboardingValidationKeySecretTemplate.DeepCopy(), timeout, interval)
+					}
```
Comment (nb-ohad) [inline controllers/managedocs_controller_test.go:695]: I think we should validate that before validating the storage cluster

File: controllers/managedocs_controller_test.go [hunk 9]
```diff
@@ -640,44 +646,69 @@ var _ = Describe("ManagedOCS controller", func() {
...
-					By("Creating an alertmanager resource")
-					utils.WaitForResource(k8sClient, ctx, amTemplate.DeepCopy(), timeout, interval)
+					Eventually(func() bool {
+						secret := onboardingValidationKeySecretTemplate.DeepCopy()
+						err := k8sClient.Get(ctx, utils.GetResourceKey(secret), secret)
+						if err != nil {
+							return false
+						}
+						return secret.Data != nil && string(secret.Data["key"]) == "-----BEGIN PUBLIC KEY-----\ntest-validation-key\n-----END PUBLIC KEY-----"
+					}, timeout, interval).Should(BeTrue())
```
Comment (nb-ohad) [inline controllers/managedocs_controller_test.go:711]: Why not use the same pattern as in other tests? (`utils.WaitForResource`)

**Review by nb-ohad — CHANGES_REQUESTED** [review]

**Review by nb-ohad — APPROVED** [review]

**Comment by nb-ohad** [comment]
@bindrad Please resolve the conflict so I can merge

**Comment by bindrad** [comment]
@nb-ohad Resolved merge conflicts
